### DISCUSSION TITLE
Controlling the sensitive value for list items

### DISCUSSION THREAD

**Discussion by pieterdd** [Relm4/Relm4#627] [Q&A] [answered]
I'd like to apply a data binding to all `RelmListItem`s in a `TypedListView`. Specifically, I'd like to control the underlying `ListItem`'s `sensitive` value via an expression that relies on the state of the model item. In gtk4-rs, the idiomatic way to disable certain items within a list appears to be to call `a_list_item.set_sensitive(false)` and make sure your `connect_activate` filters out events from the disabled items.

I know `RelmListItem::bind` has a reference to both my data model and the `ListItem` where the `sensitive` value should be applied. What is the recommended way to bind an expression to `sensitive`? I know `BoolBinding` exists, but in my case the expression is of the form `condition_1 && condition_2`.

**Comment by AaronErhardt** [comment]
I'd say you have to evaluate the expression and then update the bool binding. Or is there something that prevents you from doing that?

**Reply by pieterdd** [reply]
Wouldn't that require being notified whenever one of the two variables that make up the expression changes?

One of the variables is an enum wrapped in `SharedState`, the other is a boolean attribute from the model instance that's represented by the `ListItem`. So the expression is something along the lines of `GLOBAL_ENUM_STATE.read() == GlobalEnum::Foo && model.is_a_foo == true`. Usually a `#[watch]` would take care of the heavy lifting but it appears this is not an option here.

**Reply by AaronErhardt** [reply, marked as answer]
> Wouldn't that require being notified whenever one of the two variables that make up the expression changes?

Yes, that is certainly the case. I don't know if there's a better solution that updating each list item because every item has to be notified. Bindings are not supposed to handle expressions and that would not change the fact that you have to run some code for each list item, it would just be hidden. So in the end an hypothetical "expression binding" would just make the code a little bit cleaner.

**Reply by pieterdd** [reply]
Noted. I suppose I could instrument my data model by replacing the bool with BoolBinding and subscribing to the binding via `BoolBinding::connect_value_notify`, and use `SharedState::subscribe` for the global variable. Thanks for the guidance Aaron!

**Reply by pieterdd** [reply]
Ah dangit, it appears that `RelmListItem` doesn't have a sender I can pass to `SharedState::subscribe` :thinking:

**Reply by AaronErhardt** [reply]
You can use the `Sender` of your component for that probably (I think you just need one message and then update all items from your component). Or you [create new channels](https://docs.rs/relm4/latest/relm4/fn.channel.html) and handle the messages yourself (maybe using `relm4::spawn_local`).

**Reply by pieterdd** [reply]
Sounds interesting. Can the parent component pass on its sender to the `RelmListItem`? Can I simply pass a fresh instance of `relm4::channel:channel()` to the subscribe call within the `RelmListItem` instance? I'm not too familiar with the inner workings of Relm4 yet.

**Reply by pieterdd** [reply]
For people finding this thread in the future, something along these lines worked for me:

```rust
let (tx, rx) = relm4::channel::<TypeOfTheSharedStateVariable>();
relm4::spawn_local(async move {
    loop {
	match rx.recv().await {
	    Some(value) => {
	    	// ...
	    },
	    None => break,
	}
    }
});
SOME_SHARED_STATE.subscribe(&tx, |value| (*value).clone());
```
